### DISCUSSION TITLE
Support for non-circular (polygonal) zones.

### DISCUSSION THREAD

**Discussion by veleek** [home-assistant/frontend#16465] [Other feature requests]
I use zones to trigger a bunch of automations  I have one automation that I'd like to trigger as I'm getting close to home, so I have an additional passive "Close to Home" zone.  Unfortunately, due to the location of my house if I create a regular circular zone, there are a few main roads that I'll drive on along the way that result in certain automations triggering on the way away from the home.

![image](https://user-images.githubusercontent.com/1055021/236731980-7fa3715d-c4f1-4a37-b7d1-c9d839895d2f.png)
*Example: green area is trigger area I care about, red areas are main roads that can be "re-entered" after leaving*

In order to support this, I think it would be useful to add support for non-circular zones.  The easiest most flexible option is probably just to add support for polygonal zones, and it's already very well supported by Leaflet so it shouldn't be too hard.

I've started implementing this, but I thought it would be worthwhile to figure out if it's something that has been considered before and explicitly avoided for some reason.  @bramkragten, you may have some comments as the last person to touch a bunch of the pieces that I've been looking at.  I know that I could potentially workaround this with a series of smaller circular zones.  This specific case is just the initial driver for looking at this, I think it'd be useful functionality to have in general.

Two major components:
* Adding support for zones to have zones of different types
  * I'm also considering adding a MultiZone which is just a zone that "wraps" one or more other zones as an easy way to combine zones into a single larger zone (e.g. "Entry Gates" to have triggers that can more easily work without needing to add remove zones in multiple places.
* Updating UI to enable rendering and creating of polygonal zones.  This isn't too hard in the Add Zone page, but could make things a bit more complicated in some other places (e.g. Home zone on initial setup) so it may be better to exclude them from there.  

Potential issues:
* Performance, checking circular intersections is cheap, polygonal stuff is slightly more complicated, but for the majority of users who have a smaller number of zones maybe it's not that much.
* The UI to make it clear to users how to create and edit the polygonal zones is non-trivial.  Adding/removing points from a list is hard when those points all refer to essentially identical lat/lon values.  
  * Can we help this by just numbering the points in the polygon on the map?

**Comment by h8hr1pzdo3rmiqv5** [comment]
Upvoted.
Here's a feature request.
https://community.home-assistant.io/t/polygon-tracker-zones/26620
